GAWB-1174 estimate storage cost

breilly2 opened the pull request:
The first commit implements the functionality and includes a test that depends on the current price list. The remaining commits include clean-up and introducing a `MockGoogleServicesDAO` so that the test won't break when storage prices change.

This in only the orchestration layer. I have not made UI changes yet.

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
- [x] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [x] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [x] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [x] Tell the tech lead (TL) that the PR exists if they wants to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [x] **LR** sign off
- [ ] **Product Owner** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

breilly2 committed:
Added storageCostEstimate service for workspaces.

breilly2 committed:
Updated API to accept an implicit WithAccessToken instead of UserInfo.

breilly2 committed:
Added GoogleServicesDAO trait to allow for mocking in tests.
Made HttpGoogleServicesDAO methods private if only called internally. Added explicit result types to all other HttpGoogleServicesDAO methods.
Added MockGoogleServicesDAO and for injection into WorkspaceService for testing.

breilly2 committed:
Changed to a more appropriate exception.

breilly2 commented on src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala:
```diff
@@ -70,6 +73,19 @@ class WorkspaceService(protected val argUserToken: WithAccessToken, val rawlsDAO
+      case Seq(usage: RawlsBucketUsageResponse, priceList: GooglePriceList) =>
+        val rate = priceList.prices.cpBigstoreStorage.us
+        val estimate: BigDecimal = BigDecimal(usage.usageInBytes) / 1000000000 * rate
+        RequestComplete(WorkspaceStorageCostEstimate(f"$$$estimate%.2f"))
```
Choosing to fully format the dollar value server-side to avoid having to do any rounding/formatting client-side. Also, the price list currently only gives us US dollars, so there are no other formatting options.

breilly2 commented on src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala:
```diff
@@ -39,10 +40,10 @@ object WorkspaceService {
 }
 
-class WorkspaceService(protected val argUserToken: WithAccessToken, val rawlsDAO: RawlsDAO, val thurloeDAO: ThurloeDAO) extends Actor with AttributeSupport with TSVFileSupport {
+class WorkspaceService(protected val argUserToken: WithAccessToken, val rawlsDAO: RawlsDAO, val thurloeDAO: ThurloeDAO, val googleServicesDAO: GoogleServicesDAO) extends Actor with AttributeSupport with TSVFileSupport {
```
This is the only place where I made changes to inject a mock `GoogleServicesDAO` for tests. Other places that currently use `HttpGoogleServicesDAO` directly are traits which would require deeper refactoring to allow for injection of a mock. I'd like to leave that until it is actually needed. Let me know if you feel strongly that the refactoring should be done now.

dvoet commented on src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -44,7 +44,7 @@ object GooglePriceListJsonProtocol extends DefaultJsonProtocol {
 import org.broadinstitute.dsde.firecloud.dataaccess.GooglePriceListJsonProtocol._
 
-object HttpGoogleServicesDAO extends FireCloudRequestBuilding {
+object HttpGoogleServicesDAO extends GoogleServicesDAO with FireCloudRequestBuilding {
```
this probably should not be an object but that could be for another day

dvoet commented on src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/RawlsDAO.scala:
```diff
@@ -17,12 +17,15 @@ trait RawlsDAO extends LazyLogging {
   lazy val rawlsCuratorUrl = FireCloudConfig.Rawls.authUrl + "/user/role/curator"
   lazy val rawlsAdminWorkspaces = FireCloudConfig.Rawls.authUrl + "/admin/workspaces?attributeName=library:published&valueBoolean=true"
   lazy val rawlsWorkspaceACLUrl = FireCloudConfig.Rawls.workspacesUrl + "/%s/%s/acl"
+  lazy val rawlsBucketUsageUrl = FireCloudConfig.Rawls.workspacesUrl + "/%s/%s/bucketUsage"
```
I prefer the style of the following next line of code. Use string interpolation instead of a format string with a later call to .format. It combines the string definition and the function into 1 line, removes a string that is otherwise useless and maybe confusing, and allows us to not have a getter function.

breilly2 replied:
I like that better, too. I just followed the wrong pattern because it was the more prevalent one. I'll look into updating them all to avoid the same mistake in the future.

breilly2 replied:
The only other one here that uses `format` is `rawlsWorkspaceACLUrl`. However, it's not currently covered by any tests. They'd need to be integration tests which firecloud-orchestration currently does not have and I am not up to setting that up right now. I'll fix mine, but unfortunately I'm going to have to leave `rawlsWorkspaceACLUrl` as is.

dvoet commented on src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala:
```diff
@@ -70,6 +73,19 @@ class WorkspaceService(protected val argUserToken: WithAccessToken, val rawlsDAO
   }
 
+  def getStorageCostEstimate(workspaceNamespace: String, workspaceName: String) = {
+    Future.sequence(Seq(
```
instead of Future.sequence I think you want zip, ```rawlsDAO.getBucketUsage(workspaceNamespace, workspaceName).zip(googleServicesDAO.fetchPriceList)```

I am a little concerned about type erasure in the code as it stands.

breilly2 replied:
Yes, that makes sense. I didn't like having to explicitly specify the types. I guess maybe `Future.sequence` works better when all the futures are of the same type?

dvoet commented on src/main/scala/org/broadinstitute/dsde/firecloud/webservice/WorkspaceApiService.scala:
```diff
@@ -159,6 +159,13 @@ trait WorkspaceApiService extends HttpService with FireCloudRequestBuilding
+                perRequest(requestContext,
+                  WorkspaceService.props(workspaceServiceConstructor, userInfo),
+                  WorkspaceService.GetStorageCostEstimate(workspaceNamespace, workspaceName))
+              }
```
can we expose a pass through that is the storage in bytes as well? Seems useful from a general api perspective. Or can we include the size in the cost estimate response?

breilly2 replied:
I'm a little hesitant to add something we don't currently need (more to test and maintain), but the passthrough is simple enough so I will add that.

breilly2 committed:
WorkspaceApiServiceSpec refactoring (no JIRA, though related to work for GAWB-1174) (#325)

* Refactored WorkspaceApiServiceSpec so that mock behavior is closer to the test that it is needed for.
Improved WorkspaceApiServiceSpec to test what should be rather than what should not be.

* JSON format for List[WorkspaceACLUpdate]

* Removed unnecessary explicit listFormat for WorkspaceACLUpdate.

breilly2 committed:
Added passthrough for Rawls bucketUsage service.
Changed Future.sequence to Future.zip to avoid type erasure.

breilly2 committed:
Fixed fallout from rebase.

breilly2 commented:
Updated with changes based on all feedback and rebased onto develop as of this morning.

MatthewBemis commented:
:+1:

dvoet reviewed (approved):
👍
